src/main/java/graphql/execution/DeferSupport.java

```diff
@@ -0,0 +1,113 @@
+package graphql.execution;
+
+import graphql.Directives;
+import graphql.ExecutionResult;
+import graphql.ExecutionResultImpl;
+import graphql.GraphQLError;
+import graphql.Internal;
+import graphql.language.Field;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+/**
+ * This provides support for @defer directives on fields that mean that results will be sent AFTER
+ * the main result is sent via a Publisher stream.
+ */
+@Internal
+public class DeferSupport implements Publisher<ExecutionResult> {
+
+    private final Deque<Supplier<CompletableFuture<ExecutionResult>>> deferredCalls = new ConcurrentLinkedDeque<>();
+    private final List<GraphQLError> errorsEncountered = new ArrayList<>();
+    private final AtomicBoolean deferDetected = new AtomicBoolean(false);
+
+    public boolean checkForDeferDirective(List<Field> currentField) {
+        for (Field field : currentField) {
+            if (field.getDirective(Directives.DeferDirective.getName()) != null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
```

Comment:
So whats intresting about this PR is that its giving off a non spec compliant ReactiveStream

https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.2/README.md#specification

That is the rx stream you get does not follow all the capabilities a true stream should have .  This is a great read on that

https://medium.com/@olehdokuka/mastering-own-reactive-streams-implementation-part-1-publisher-e8eaf928a78c

You would need to have Rxjava / reactor.io to get a truly 100% reactive Publisher.  We get that in subscriptionsbecause we delegate to some one else to give is the Publisher

But.... here is the thing.  The shape is perfect for sending a stream of results later.

That is the onNext() callback shape and the subscribe() call is perfect.

The `subscribe()` call is when the down stream subscriber has connected to the publisher.

This is the perfect time to start the running of the deferred calls.  If the caller has the `Publisher` then it must be ready to get further results.  As soon as they call subscribe we can start executing the queue of deferred fields.

As each one finishes, we send it down the pipe as a series of `onNext` calls.

This is why the Publisher shae is so great even though its not a 100% reactive generic implementation.

So the subscription object we give them is empty.  We wont have any back pressure on the deferred fields say (they cant ask for 2 and then ask for 2 more etc..)

We could invent a way for some to SPI plugin an actual Publisher say - but I am unconvinced yet that its needed.  I mean how many deferred fields will you have?  10s, 100s or 10,000?  Its the low number not a high number where you need back pressure